Modify gci option in --extract flag

dexter-1 opened the pull request:
By default, the --extract=gci/FAMILY flag will look at COS images in
the GCP project "container-vm-image-staging" and use the GCS bucket
"gs://container-vm-image-staging/k8s-version-map" to map the image name
to k8s version to be extracted. The COS team wants to run e2e tests on images found in other GCP projects. These values are hard-coded, and it
would be ideal to pass that information to the --extract flag instead.

dexter-1 commented:
/assign @krzyzacy

krzyzacy commented:
Can you assign one of your teammate to review, and also add a test for this?

kubetest is mostly in maintenance mode, but as long as your team wants to maintain this logic I'm happy to take this in, looks like nobody else is using this code path.

krzyzacy commented:
/ok-to-test

wonderfly commented:
/assign @krzyzacy

dexter-1 commented:
/cc @wonderfly

dexter-1 commented:
Thanks @krzyzacy for taking a look at this PR so quickly. 

I added my teammate, @wonderfly to review the PR. I also added unit tests for the changes I made. Our team is fine maintaining this logic

bsdnet commented on kubetest/extract_k8s.go:
```diff
@@ -346,9 +346,30 @@ func setReleaseFromHTTP(prefix, suffix string, getSrc bool) error {
-	p := "container-vm-image-staging"
-	b, err := control.Output(exec.Command("gcloud", "compute", "images", "describe-from-family", family, fmt.Sprintf("--project=%v", p), "--format=value(name)"))
+var parseGciExtractOption = func(option string) (string, map[string]string) {
+	arr := strings.Split(option, "?")
```
s/arr/tokens/?

dexter-1 replied:
Done.

bsdnet commented on kubetest/extract_k8s.go:
```diff
@@ -346,9 +346,30 @@ func setReleaseFromHTTP(prefix, suffix string, getSrc bool) error {
+		params := strings.Split(arr[1], ":")
+		for _, param := range params {
+			kv := strings.Split(param, "=")
+			paramsMap[kv[0]] = kv[1]
```
This will crash if len(kv) < 2

dexter-1 replied:
kv will always be length two since this represents a kv pair, and the regex in line 78 enforces this.

bsdnet commented on kubetest/extract_k8s.go:
```diff
@@ -346,9 +346,30 @@ func setReleaseFromHTTP(prefix, suffix string, getSrc bool) error {
+	return control.Output(exec.Command("gcloud", "compute", "images", "describe-from-family", family, fmt.Sprintf("--project=%v", project), "--format=value(name)"))
+}
+
+func setupGciVars(family string, p string) (string, error) {
```
either family string, project string 
or f, p string

dexter-1 replied:
Change to f, p string.

bsdnet commented on kubetest/extract_k8s.go:
```diff
@@ -346,9 +346,30 @@ func setReleaseFromHTTP(prefix, suffix string, getSrc bool) error {
+	paramsMap := map[string]string{
+		// default values
+		"project":            "container-vm-image-staging",
+		"version-map-bucket": "container-vm-image-staging",
```
s/version-map-bucket/k8s-map-bucket/g

dexter-1 replied:
Done.

bsdnet commented on kubetest/extract_k8s.go:
```diff
@@ -435,12 +456,15 @@ func (e extractStrategy) Extract(project, zone, region string, extractSrc bool)
 		return getKube(fmt.Sprintf("file://%s", url), release, extractSrc)
 	case gci, gciCi:
-		if i, err := setupGciVars(e.option); err != nil {
+		family, gciExtractParams := parseGciExtractOption(e.option)
```
s/gciExtractParams/extraParams/g

dexter-1 replied:
imo, I think gciExtractParams more clearly describes this variable, as these "params" only belong to the --extract=gci option.

bsdnet commented on kubetest/extract_k8s.go:
```diff
@@ -435,12 +456,15 @@ func (e extractStrategy) Extract(project, zone, region string, extractSrc bool)
 			return setReleaseFromGcs("kubernetes-release-dev/ci", e.ciVersion, extractSrc)
 		} else {
-			return setReleaseFromGci(i, extractSrc)
+			return setReleaseFromGci(i, versionMapBucket, extractSrc)
```
s/versionMapBucket/k8sMapBucket/g

dexter-1 replied:
Done.

wonderfly commented:
/lgtm
